feat: Share Serial Nos between unique Items

marination opened the pull request:
`no-docs` (temporary)

## Case:
- Serial Nos are purchased from Supplier and must be maintained in the system as per supplier
- A company buys different items from a supplier, who (supplier) starts all their serialised items with 1 and then autoincrements. This way item A can have serial “3” and item B can also have serial no “3”.
- Additionally: A serial no is also something that a supplier/ manufacturer can choose freely.

## Solution:
- Make Serial Nos unique only on Item level (Item A cannot have Serial No "3" twice)
- This means Item A and B can have the same Serial No, but different Serial No document names
  <img width="1328" alt="Screenshot 2023-09-29 at 4 50 18 PM" src="https://github.com/frappe/erpnext/assets/25857446/13a53bb6-ecfc-4862-a9f8-534ef85deb79">

## Flow:
- Serialised Item A receives some Serial Nos via a Purchase Receipt. These are auto created using a naming series. **SN-0001** is one of them
- We want to register a Serial No for Serialised Item B, which also has the same Serial No
- We can do one of the following:
    - Create a new Serial No (manual or data import) > Use it in the new Purchase Receipt's popup, by scanning or entering it in the table
    - In the new Purchase Receipt popup, upload a CSV of the Serial No(s).
    - Create a new Serial and Batch Bundle > click on "Make Serial Nos" in the form > insert/upload the Serial No(s)
- Now we should have a Serial and Batch Bundle created that includes a Serial No  with **ID**: **SN-0001-1** and **Serial No**: **SN-0001**

All of the above methods handle naming clashes via the `autoname` function in the Serial No. 
  

> **Note:** This does **not** work for naming series (auto naming) as the last count is maintained series wise. So if we have a series `SN-.####` shared by two items, the insertion of the Item-A creates SN-0001 (series count is 1) and then the following insertion of Item-B will naturally create SN-0002.

> For now only manual addition of serial no.s through various modes is supported for reusability

**ToDo:**

- [x] Link formatting for Serial No in tables and reports
- [x] Tests

marination committed:
feat: Share Serial Nos between unique Items

- Amend Serial No name on Insertion of Serial No if it has a duplicate in another item
- `get_serial_nos` must accept item_code since serial not can be same for different items.
- SN text fields contain the Serial No not the docname
- Removed `get_serial_nos_data`

marination committed:
fix: Optional `item_code` in `get_serial_nos`

- If duplicate serial nos are allowed, `get_serial_nos` requires item code, throw error.
- Else, `item_code` is optional for backwards compatibility

marination committed:
chore: Adjust `make_serial_nos`

- On uploading new Serial Nos via CSV, make new SNs and insert SN Name in serial batch bundle
- Remove unused args & pass item_code to `get_available_serial_nos`

marination committed:
fix: Wrong fieldname passed to `get_serial_nos`

marination committed:
fix: Link formatting for Serial No

- Add "serial_no_name" hidden field and use it to format Serial No link field
- Format Link fields in documents, child tables (Serial and Batch Entry) and report link columns

marination committed:
fix: Format SN link in "Pick Serial/Batch No" pop up

marination committed:
test: Insertion of Duplicate SN in another Item via PR

- Insert two SNs, one via PR, the other via S&B Bundle(via text popup or csv)

marination committed:
fix: Throw UI error on scanning invalid Serial No in inward popup

marination committed:
fix: Avoid `args` usage

marination committed:
fix: Use (SN, SN Name) as key to check if SN is reserved

- Set Serial No name in SRE table
- Use (SN, SN Name) as key to check if SN is reserved

marination commented:
@rohitwaghchaure Whenever you can please take a look. Would love to get your review :)

marination commented:
TODO: Add more tests for stock transfer, stock reservation, delivery note

bosue commented:
True, the Serial Number can’t be unique across all items.
However, if it isn’t unique, so we need a separate unique ID, then I think it’s less confusing if that unique ID is a random hash.
We certainly don‘t want users to confuse the two. Also, unlike with, say, amended submittable documents, there may be nothing two accidentally identical S/Ns that are referring to different items have in common.

Still we may want an additional provision allowing to enforce unique serial numbers over a particular Item Group, as for some items out in the world, general standards for S/N exist. But that‘s matter for a separate ticket.

barredterra commented:
@bosue this PR is supposed to get backported all the way to v14, so we can't just change the naming scheme. This is a completely optional approach, for users who are aware of the implications. Regarding future versions I agree with you, but that will have to be a separate PR.

bosue commented on erpnext/stock/doctype/stock_settings/stock_settings.json:
```diff
@@ -420,6 +421,13 @@
+  },
+  {
+   "default": "0",
+   "description": "Allows duplicacy across separate items. Maintains uniqeness within an item.",
```
- In standard English, unfortunately there only seems to be the word “duplication,” not “[duplicacy](https://en.wiktionary.org/wiki/duplicacy)”, though I personally like it.
- “uniq**u**eness”
- Also, we need to explain what may be duplicated.

Propose: “Allow duplicate serial numbers across different Items. Maintain uniqueness per Item.”

marination replied:
That does sound better. Do leave a direct code review next time, quicker to commit and also co-author you

bosue replied:
Sry, I would if I could. But the Code review feature isn't available to me as a normal contributor... 💁🏻‍♂️

bosue commented on erpnext/stock/doctype/stock_settings/stock_settings.json:
```diff
@@ -420,6 +421,13 @@
+   "description": "Allows duplicacy across separate items. Maintains uniqeness within an item.",
+   "fieldname": "allow_duplicate_serial_nos",
+   "fieldtype": "Check",
+   "label": "Allow Duplicate Serial Nos Between Items"
```
“Across” should be better here, too.

bosue reviewed (commented):
Just tidbits.

If we already know that at a later point – preferably soon – we want to allow Uniqueness per Item Group as a third option, we might want to use a select rightaway so we're flexible. But it‘s perfectly fine to keep it simple now.

marination commented:
@barredterra To Test:
- Enable the setting
- Create two serialised items. Let the first item be autonamed and create serial nos for it (Purchase Receipt) (eg. SN-001, SN-002, etc.)

To insert Serial Nos for the second item, these are multiple ways to achieve the same goal :
- Use the "Download Template" > "Attach CSV" feature. Make sure the CSV has SN-002 (same as the first item) and any other SN. The expectation is that the the new SN should be created under the name SN-002-1
   <img width="300" alt="Screenshot 2023-11-27 at 5 58 48 PM" src="https://github.com/frappe/erpnext/assets/25857446/cf863658-4918-4cd0-aa1d-3bafe665ffbf">
- Manually create a Serial No from the serial no. list view and select it in the popup above. A duplicate Serial No of another item should work just fine
- Use Data Import to insert a duplicate Serial No record
- In "Serial and Batch Bundle", fill in the basic details, click on "Make Serial Nos", try inserting a duplicate using the text field and CSV. 

To move the serial no:
- In a Delivery Note, use the **Scan Serial No** field in the Serial No popup. The expectation is that if **SN-002** is scanned for the second item, the table in the popup should be populated with **SN-002-1**
- The same can be checked for a stock entry where the Serial No is transferred from one warehouse to another
- The same check can be done for a Serial and Batch Item, along with the right Serial No ID, the right batch no. will be populated in the popup table too

Stock Reservation:
- Enable Stock Reservation in Stock Settings https://docs.erpnext.com/docs/user/manual/en/stock-reservation and also enable auto reserve serial and batch nos
- Create a Sales Order for the second Serial No (with duplicate) > submit > Reserve stock
- Check that the created stock reservation entry has our duplicate serial no with the right name that links to the right record

You can check the PR description where Serial No. link formatting is also fixed in this PR for a duplicate case.

marination committed:
refactor: `update_serial_batch_no` (serial no part)

barredterra commented on erpnext/stock/doctype/serial_no/serial_no.py:
```diff
@@ -203,14 +252,15 @@ def auto_fetch_serial_number(
 	else:
 		exclude_sr_nos = safe_json_loads(exclude_sr_nos)
-		exclude_sr_nos = get_serial_nos(clean_serial_no_string("\n".join(exclude_sr_nos)))
+		exclude_sr_nos = get_serial_nos(clean_serial_no_string("\n".join(exclude_sr_nos), item_code))
```
I can't make much sense of this line.

I suspect the original intention was to indirectly call `.upper().strip()` on each serial number. That could also be achieved by the simpler

```suggestion
		exclude_sr_nos = get_serial_nos("\n".join(exclude_sr_nos))
```

, since

```python
serial_numbers = "\n".join([" a", "b "]
(
	get_serial_nos(clean_serial_no_string(serial_numbers))
	== get_serial_nos(serial_numbers)
	== ["A", "B"]
)
```

If we add the `item_code` parameter, the behavior differs in that we'd get `None` for missing serial numbers instead of the cleaned up string.

marination replied:
`get_serial_nos` is mainly used to handle legacy data where the serial no text field is used.
Anyway, it will return the serial no back as is if it cannot be found to keep it consistent with older behaviour.
The retrieval of serial no name is added as a defensive approach.

barredterra commented on erpnext/public/js/utils.js:
```diff
@@ -937,6 +937,17 @@ frappe.form.link_formatters['Project'] = function(value, doc) {
+	} else {
+		return value;
+	}
+}
```
I'm not sure if this is helpful. The system-internal name, i. e. the serial number with postfix, is probably not very important for the user. And when it is, it's just one click away.

Also,

<a href="#AAA-3">AAA-3: AAA</a>
<a href="#BBB-1">BBB-1: BBB</a>
<a href="#CCC">CCC</a>
<a href="#DDD-8">DDD-8: DDD</a>

looks much more noisy than

<a href="#AAA-3">AAA</a>
<a href="#BBB-1">BBB</a>
<a href="#CCC">CCC</a>
<a href="#DDD-8">DDD</a>.

Why don't we just enable _Show Title in Link Fields_ in the **Serial No** doctype?

barredterra commented on erpnext/public/js/utils/serial_no_batch_selector.js:
```diff
@@ -352,44 +361,35 @@ erpnext.SerialBatchPackageSelector = class SerialNoBatchBundleUpdate {
-					serial_no: scan_serial_no
-				});
+			frappe.db.get_value(
+				"Serial No",
+				{"serial_no": scan_serial_no, "item_code": this.item.item_code},
+				this.item.has_batch_no ? ["name", "batch_no"] : ["name"],
+				(res) => {
+					if (!res || res.exc) {
+						frappe.throw(
+							__(
+								"Serial No {0} does not belong to Item {1}",
```
Stick with the single quotes, for consistent style with the surrounding code?

Or maybe just bulk format this file in a separate style PR, after this one is merged.

marination replied:
will do it in a separate PR, theres a lot of double quotes elsewhere in the file

barredterra commented on erpnext/stock/doctype/serial_no/serial_no.py:
```diff
@@ -64,6 +64,37 @@ def __init__(self, *args, **kwargs):
 		self.via_stock_ledger = False
 
+	def autoname(self):
+		self.name = self.serial_no
+
+		if frappe.db.get_single_value("Stock Settings", "allow_duplicate_serial_nos"):
+			self.validate_duplicate_in_same_item()
+			self.name = self.autoname_duplicate_serial_no()
```
```suggestion
		if frappe.db.get_single_value("Stock Settings", "allow_duplicate_serial_nos"):
			self.validate_duplicate_in_same_item()
			self.name = self.autoname_duplicate_serial_no()
		else:
			self.name = self.serial_no
```

A bit more readable, IMO. (I don't have to worry about whether the default name could influence the duplicate name.)

marination committed:
Revert "fix: Link formatting for Serial No"

This reverts commit 789a91f2653a8573459c4a336ac371cd516192cf.

marination committed:
Revert "fix: Format SN link in "Pick Serial/Batch No" pop up"

This reverts commit 4cddf09d57244af9c95f0adb42b4a3b3b310e687.

marination committed:
fix: Show title in link fields for SN & `get_serial_nos` always returns a SN

- `get_serial_nos` must return input SN if SN does not exist (consistent with old behaviour)
- Show title in link fields for SN as link formatting

marination committed:
Merge branch 'develop' into shared-serial-nos

marination commented:
@rohitwaghchaure please give this a look for any potential breaking issues ?

barredterra committed:
Merge branch 'develop' into shared-serial-nos

barredterra committed:
fix: don't query nonexistent field

`serial_no` is a link field, so `serial_no_name` seems redundant.

barredterra commented:
Linter failure is unrelated; fixed in https://github.com/frappe/erpnext/pull/39511

barredterra committed:
Merge branch 'develop' into shared-serial-nos

barredterra committed:
Merge branch 'develop' into shared-serial-nos

barredterra commented:
@rohitwaghchaure since you are the expert on this topic, we would really appreciate your review and guidance on this PR.

barredterra committed:
fix: SN filter in report "Cost of Poor Quality"

barredterra commented:
Manual tests completed

Before – after comparison of reports:

- [x] Cost of poor Quality (no data)
- [x] Incorrect Serial No Valuation (no data)
- [x] Serial No Ledger
- [x] Stock Ledger
- [x] Stock Qty vs Serial No Count

Sales and Purchase Cycle with one serialized and one non-serialized item. Setting "Allow Duplicate Serial Nos Across Items" disabled.

- [x] Purchase Order
- [x] Purchase Receipt
- [x] Stock Entry (Type: "Material Transfer")
- [x] Stock Reconciliation
- [x] Sales Order
- [ ] Stock Reservation
- [ ] Subcontracting
- [x] Delivery Note

Sales and Purchase Cycle with one serialized and one non-serialized item. Setting "Allow Duplicate Serial Nos Across Items" enabled.

- [x] Purchase Order
- [x] Purchase Receipt
- [x] Stock Entry (Type: "Material Transfer")
- [x] Stock Reconciliation
- [x] Sales Order
- [ ] Stock Reservation
- [ ] Subcontracting
- [x] Delivery Note

barredterra committed:
fix: pass item_code to get_serial_nos

barredterra committed:
fix: use item code when checking existence

barredterra committed:
fix: make_serial_no_if_not_exists

rohitwaghchaure commented:
Found few failed cases during the testing

- User Serial / Batch fields feature not working
  - Created same serial number with different items first
  - While making the purchase receipt, enabled "Use Serial / Batch Fields" checkbox in the line item.
  - Enter the Serial Number created at step 1 and submitted
  - System has created new serial number with has value, also system has ignored "Serial Number Series" from the item master while creating new serial no (primary key should has value not the serial no field's value)
- Getting below error while making the stock reconciliation with User Serial / Batch fields
<img width="674" alt="image" src="https://github.com/frappe/erpnext/assets/8780500/1295fcf6-f147-4846-9209-1942df25121d">

marination commented:
ToDo and Concerns:
- <s>Test Serial No Series, if autoname obstructs serial no series naming </s>
- text field should have serial no (and not name) and name should be handled on the backend
- Case: Scanning serial no does not know which item the serial no belongs to, same with POS
- Serial Nos with naming series are not handled. (We enable this on an item level, where the name is a hash value, but naming series on the SN of the item)
- Db level unique key constraint
- DX

PatrickDEissler commented:
Comment against inactivity.

ankush commented on erpnext/stock/doctype/serial_no/serial_no.py:
```diff
@@ -64,6 +64,37 @@ def __init__(self, *args, **kwargs):
+
+	def validate_duplicate_in_same_item(self):
+		"""Check to make sure duplicates are not allowed in the same item."""
+		if frappe.db.exists("Serial No", {"serial_no": self.serial_no, "item_code": self.item_code}):
+			frappe.throw(
+				msg=_("Serial No {0} already exists for Item {1}").format(
+					frappe.bold(self.serial_no), frappe.bold(self.item_code)
+				),
+				title=_("Duplicate"),
+			)
```
While validation is fine, this should also be a DB level unique constraint using `db.add_unique` we have had many such problem with `Bin` where ORM level validations are not sufficient.
